src/Factory.ts

```diff
@@ -1,17 +1,75 @@
 import { Node } from './Node';
+import { GetSet } from './types';
 import { Util } from './Util';
 import { getComponentValidator } from './Validators';
 
-var GET = 'get',
-  SET = 'set';
+const GET = 'get';
+const SET = 'set';
+
+/**
+ * Enforces that a type is a string.
+ */
+type EnforceString<T> = T extends string ? T : never;
+
+/**
+ * Represents a class.
+ */
+type Constructor = abstract new (...args: any) => any;
+
+/**
+ * An attribute of an instance of the provided class. Attributes names be strings.
+ */
+type Attr<T extends Constructor> = EnforceString<keyof InstanceType<T>>;
+
+/**
+ * A function that is called after a setter is called.
+ */
+type AfterFunc<T extends Constructor> = (this: InstanceType<T>) => void;
+
+/**
+ * Extracts the type of a GetSet.
+ */
+type ExtractGetSet<T> = T extends GetSet<infer U, any> ? U : never;
+
+/**
+ * Extracts the type of a GetSet class attribute.
+ */
+type Value<T extends Constructor, U extends Attr<T>> = ExtractGetSet<
+  InstanceType<T>[U]
+>;
+
+/**
+ * A function that validates a value.
+ */
+type ValidatorFunc<T> = (val: ExtractGetSet<T>, attr: string) => T;
+
+/**
+ * Extracts the "components" (keys) of a GetSet value. The value must be an object.
+ */
+type ExtractComponents<T extends Constructor, U extends Attr<T>> = Value<
+  T,
+  U
+> extends Record<string, any>
+  ? EnforceString<keyof Value<T, U>>[]
+  : never;
 
 export const Factory = {
-  addGetterSetter(constructor, attr, def?, validator?, after?) {
+  addGetterSetter<T extends Constructor, U extends Attr<T>>(
+    constructor: T,
+    attr: U,
```

Comment:
We can make this a bit safer with a type to constrain `attr` to be a `GetSet`.

```ts
type EnforceGetSet<
  T extends Constructor,
  U extends Attr<T>
> = InstanceType<T>[U] extends GetSet<any, any> ? U : never;
```

Then, type `attr` as `attr: EnforceGetSet<T, U>`.

However, this causes a problem in `Canvas.ts` at https://github.com/psychedelicious/konva/blob/9c41a6eed1bba46e80ce282903ac86bb11493e84/src/Canvas.ts#L171